src/domain/auth-rebac/domain/src/repos/rebac_repository.rs

```diff
@@ -0,0 +1,494 @@
+        PropertyName::Dataset(value)
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum AccountPropertyName {
+    // TBA
+}
+
+impl From<AccountPropertyName> for PropertyName {
+    fn from(value: AccountPropertyName) -> PropertyName {
+        PropertyName::Account(value)
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum Relation {
+    AccountDatasetReader,
+    AccountDatasetEditor,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait RebacRepository: Send + Sync {
+    // Properties
+
+    async fn set_entity_property(
+        &self,
+        entity: &Entity,
+        property: PropertyName,
+        property_value: &PropertyValue,
+    ) -> Result<(), SetEntityPropertyError>;
+
+    async fn delete_entity_property(
+        &self,
+        entity: &Entity,
+        property_name: PropertyName,
+    ) -> Result<(), DeleteEntityPropertyError>;
+
+    async fn get_entity_properties(
+        &self,
+        entity: &Entity,
+    ) -> Result<Vec<(PropertyName, PropertyValue)>, GetEntityPropertiesError>;
+
+    // Relations
+
+    async fn insert_entities_relation(
+        &self,
+        subject_entity: &Entity,
+        relationship: Relation,
+        object_entity: &Entity,
+    ) -> Result<(), InsertEntitiesRelationError>;
+
+    async fn delete_entities_relation(
+        &self,
+        subject_entity: &Entity,
+        relationship: Relation,
+        object_entity: &Entity,
+    ) -> Result<(), DeleteEntitiesRelationError>;
+
+    async fn get_subject_entity_relations(
+        &self,
+        subject_entity: &Entity,
+    ) -> Result<Vec<ObjectEntityWithRelation>, SubjectEntityRelationsError>;
+
+    async fn get_subject_entity_relations_by_object_type(
+        &self,
+        subject_entity: &Entity,
+        object_entity_type: EntityType,
+    ) -> Result<Vec<ObjectEntityWithRelation>, SubjectEntityRelationsByObjectTypeError>;
+
+    async fn get_relations_between_entities(
+        &self,
+        subject_entity: &Entity,
+        object_entity: &Entity,
+    ) -> Result<Vec<Relation>, GetRelationsBetweenEntitiesError>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Errors
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum SetEntityPropertyError {
+    #[error(transparent)]
+    Internal(InternalError),
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Error, Debug)]
+pub enum DeleteEntityPropertyError {
+    #[error(transparent)]
+    EntityNotFound(EntityNotFoundError),
+
+    #[error(transparent)]
+    PropertyNotFound(EntityPropertyNotFoundError),
```

Comment:
Make sure these errors are needed. This might be an overkill as we prefer idempotent handlers.

Reply:
> Make sure these errors are needed

For repositories, we usually use errors like "not found" and "duplicate"

> This might be an overkill as we prefer idempotent handlers.

But for `RebacService` the errors will be updated

Reply:
`UnsetEntityPropertyError` & `InsertRelationError` added -- Done